spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -607,14 +815,89 @@ private void commitIfNecessary() {
 				this.logger.debug("Committing: " + commits);
 			}
 			if (!commits.isEmpty()) {
-				if (this.syncCommits) {
-					this.consumer.commitSync(commits);
+				try {
+					if (this.syncCommits) {
+						this.consumer.commitSync(commits);
+					}
+					else {
+						this.consumer.commitAsync(commits, this.commitCallback);
+					}
 				}
-				else {
-					this.consumer.commitAsync(commits, this.commitCallback);
+				catch (WakeupException e) {
+					// ignore - not polling
 				}
 			}
 		}
+
+		private final class ListenerInvoker implements SchedulingAwareRunnable {
+
+			private volatile boolean active = true;
+
+			@Override
+			public void run() {
+				while (this.active) {
+					try {
+						ConsumerRecords<K, V> records = ListenerConsumer.this.recordsToProcess.poll(1, TimeUnit.SECONDS);
```

Comment:
... Instead of `this.listenerInvokerFuture.get(getShutdownTimeout(), TimeUnit.MILLISECONDS);`, of course.